### ISSUE TITLE
no-(s)css-rules

### ISSUE THREAD

**Issue by villesau** [kristerkari/stylelint-scss#210]

**Comment by kristerkari** [comment]
That sounds good, probably we just need to come up with a good name for the rule :)

**Comment by cainrus** [comment]
Do not forget about mixins only files.

**Comment by kristerkari** [comment]
Maybe it's a good idea to properly specify how this rule should work before starting to work on it. 

@villesau could you write something like similar than in this example: https://github.com/kristerkari/stylelint-scss/issues/103#issue-210572201 ?

**Comment by villesau** [comment]
@kristerkari I think that's not quite the same. The motivation behind this is that we have common variables.scss file, which is imported hundreds of times across the project. It mostly contains colors and some other variables. Anyhow, we had a case where some one introduced a css rule there, like:

variables.scss:
```
$color1: $blue;
$color2: $cyan;
$color3: $green5;

.my-rule {
margin: 5px
}

```

And what happened was that when variables.scss is imported hundreds of times ( `@import 'variables';` ), my-rule was multiplied into resulting css file hundreds of times, which bloated resulting css file and made dev-tools style section unusable slow. rules gets duplicated when importing, variables does not. To prevent this in future, we should be able to define comment rule per-file (only for variables.scss in this case) that it can't contain css rules like my-rule.

**Comment by kristerkari** [comment]
@villesau What I meant that could you use a similar format to specify what are the rule options, what are some examples for invalid code, and what are some examples of valid code.

For example, would you define in the plugin options that the rule only applies to a file called `variables.scss`, or would you turn on the rule inside that file?

**Comment by kristerkari** [comment]
ah, you already answered my question:

> To prevent this in future, we should be able to define comment rule per-file (only for variables.scss in this case) that it can't contain css rules like my-rule.

**Comment by chimericdream** [comment]
I just posted in #124 about this, but thought I'd put my suggestion here as well, because I think the two rules could have similar configurations. Using the convention I suggested in the other issue, the config for this rule could look something like this:

```
/* global module */
'use strict';

module.exports = {
    'extends': ['stylelint-config-recommended-scss'],
    'rules': {
        'scss/rules-filename-blacklist': [/* list of file names */]|null, // any files in the list will trigger an error if a rule is found; set to null to allow rules anywhere (default: null)
        // ...
        // the rest of your rules
    }
};
```

An added benefit to doing it this way is you wouldn't have to worry about special handling for mixins, placeholders, etc. Just specify which files are not allowed to have rule blocks, and you're done. Of course, if there is enough demand, perhaps this could be broadened to include more granular rules (e.g. mixin-only files, variable-only files, etc).

**Comment by adrianjost** [comment]
I haven't had time to release it as a plugin yet, but that's what I wrote for our codebase.
The restriction to certain files can be done in the stylelint config, by only activating this rule on those files using overwrites.

```javascript
module.exports = {
  // ...
  overrides: [
    {
      files: ["src/lib/styles/utils/**/*.scss"],
      rules: {
        "namespace/no-root-level-css": true,
      },
    },
  ],
};
```

```javascript
const stylelint = require("stylelint");
const namespace = require("../namespace");

const ruleName = namespace("no-root-level-css");

const messages = stylelint.utils.ruleMessages(ruleName, {
  noRootLevelRules:
    "Top-level CSS definitions are not allowed in this file. Consider moving them, or wrapping with an @mixin.",
  noRootLevelInclude: "Top-level @include is not allowed in this file.",
});

function rule(primaryOption, options = {}, context = {}) {
  return (root, result) => {
    if (!primaryOption) {
      return; // rule disabled
    }

    const noRootLevelInclude = (atRule) => {
      if (atRule.name !== "include") {
        return;
      }
      if (atRule.parent.type !== "root") {
        return;
      }
      stylelint.utils.report({
        message: messages.noRootLevelInclude,
        node: atRule,
        result,
        ruleName,
      });
    };

    const noRootLevelRules = (rule) => {
      if (rule.parent.type !== "root") {
        return;
      }
      stylelint.utils.report({
        message: messages.noRootLevelRules,
        node: rule,
        result,
        ruleName,
      });
    };

    root.walkRules(noRootLevelRules);
    root.walkAtRules(noRootLevelInclude);
  };
}
rule.ruleName = ruleName;
rule.messages = messages;
module.exports = rule;
```
